src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -499,19 +502,87 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
                 return Ok(res.try_into().unwrap());
             }
             Err(ResolutionFailure::NotResolved {
+                item_id,
                 module_id,
                 partial_res: None,
                 unresolved: path_str.into(),
             })
         })
     }
 
+    fn resolve_self_ty(&self, path_str: &str, ns: Namespace, item_id: ItemId) -> Option<Res> {
+        if ns != TypeNS || path_str != "Self" {
```

Comment:
Why have `ns` if it's always `TypeNS`? Likewise with `path_str`.

Reply:
Just to keep all the `Self` resolution logic in a single function.

Reply:
Also, `ns` shouldn't ideally be only `TypeNS`, `Self` in value namespace should also be resolved to constructors if we are doing things correctly.

But that also requires some more logic in ` fn resolve_with_disambiguator` (where we accumulate resolutions from all namespaces) to avoid reporting false ambiguities (similar to `is_derive_trait_collision`).
I'm not sure why other "type vs its constructors" collisions are not processed in the same place.